sql/parser/builtins.go

```diff
@@ -72,6 +75,74 @@ var builtins = map[string]builtin{
 	"upper": stringBuiltin(func(s string) (Datum, error) {
 		return DString(strings.ToUpper(s)), nil
 	}),
+
+	// TODO(XisiHuang): support the substring(str FROM x FOR y) syntax.
+	"substring": {
+		nArgs: -1,
+		fn: func(args DTuple) (Datum, error) {
+			argsNum := len(args)
+			if argsNum != 2 && argsNum != 3 {
+				return DNull, fmt.Errorf("incorrect number of arguments: %d vs %s",
+					argsNum, "2 or 3")
+			}
+
+			dstr, ok := args[0].(DString)
+			if !ok {
+				return DNull, argTypeError(args[0], dstr.Type())
+			}
+			str := string(dstr)
```

Comment:
Decoding arguments is going to be common in these builtins. I'm thinking about how to simplify this. What if we had a sort of `getArgs` method that looked like:

```
func getArgs(args DTuple, minArgs, maxArgs int, vals ...interface{}) error {
  // Check minArgs <= len(args) <= maxArgs.
  for i, v := range vals {
    switch t := v.(type) {
      case *int:
        // Require args[i].(Dint)
        x, ok := args[i].(DInt)
        if !ok {
          return argTypeError(...)
        }
        *t = x
      case *string:
       // Require args[i].(DString)
    }
  }
  return nil
}
```

And then this would be used as:

```
  var str string
  var start, length int
  if err := getArgs(args, 2, 3, &str, &start, &length); err 1= nil {
    return DNull, err
  }
  // Use str, start and length
```

Again, not for this PR, just thinking out loud.